src/libs/vehicle/mavlink/vehicle.ts

```diff
@@ -773,6 +774,68 @@ export abstract class MAVLinkVehicle<Modes> extends Vehicle.AbstractVehicle<Mode
     sendMavlinkMessage(heartbeatMessage)
   }
 
+  /**
+   * Send system time from GCS
+   */
+  sendSystemTime(): void {
+    const systemTimeMessage: Message.SystemTime = {
+      type: MAVLinkType.SYSTEM_TIME,
+      time_unix_usec: Math.floor(Date.now() * 1000),
+      time_boot_ms: Math.floor(Date.now() / 1000),
+    }
+    sendMavlinkMessage(systemTimeMessage)
+  }
+
+  /**
+   * Measure network latency to vehicle using TIMESYNC message
+   * @returns {Promise<void>}
+   */
+  async measureLatency(): Promise<void> {
+    const time_sent = Math.floor(Date.now()) * 1000
+    let receivedTimesync = false
+    let timeoutReached = false
+    let latencyUs = 0
+    const timeout = 10000
+
+    const timesyncHandler = (pack: Package): void => {
+      const timesync = pack.message as Message.Timesync
+      if (timesync.ts1 === time_sent) {
+        const time_received = Math.floor(Date.now()) * 1000
+        // we could use the tc1 value to calculate the latency, but I think that would give us the
+        // GCS -> vehicle latency, not the round trip latency.
+        latencyUs = (time_received - time_sent) / 2
+        receivedTimesync = true
+      }
+    }
+
+    const timeSyncMessage: Message.Timesync = {
+      type: MAVLinkType.TIMESYNC,
+      ts1: time_sent,
+      tc1: 0,
+      target_system: this.currentSystemId,
+      target_component: 1,
+    }
+    sendMavlinkMessage(timeSyncMessage)
+
+    const startTime = new Date()
+    this.onIncomingMAVLinkMessage.add(MAVLinkType.TIMESYNC, timesyncHandler)
```

Comment:
If this is slow/delayed, could it miss the response? I feel like if the handler is being added dynamically then it's useful to do so _before_ sending the message (especially since there's internal validation for whether the message instance being handled is the intended one).

Reply:
good catch, I remember fixing it, but apparently I didn't